src/messaging/index.ts

```diff
@@ -0,0 +1,366 @@
+/**
+ * Copyright (c) 2018-2023, BOTLabs GmbH.
+ *
+ * This source code is licensed under the BSD 4-Clause "Original" license
+ * found in the LICENSE file in the root directory of this source tree.
+ */
+
+/**
+ * KILT participants can communicate via a 1:1 messaging system.
+ *
+ * All messages are **encrypted** with the encryption keys of the involved identities.
+ * Messages are encrypted using authenticated encryption: the two parties authenticate to each other, but the message authentication provides repudiation possibilities.
+ *
+ * The [[Message]] class exposes methods to construct and verify messages.
+ *
+ * @module @kiltprotocol/messaging
+ */
+
+/**
+ * Copyright (c) 2018-2023, BOTLabs GmbH.
+ *
+ * This source code is licensed under the BSD 4-Clause "Original" license
+ * found in the LICENSE file in the root directory of this source tree.
+ */
+
+import type {
+  DecryptCallback,
+  DidResolveKey,
+  DidResourceUri,
+  EncryptCallback,
+  IEncryptedMessage,
+  IEncryptedMessageContents,
+  ICType,
+  IDelegationData,
+  IMessage,
+  MessageBody,
+} from '../types'
+import { Attestation, Claim, Credential, CType, Quote } from '@kiltprotocol/core'
+import { DataUtils, SDKErrors, UUID } from '@kiltprotocol/utils'
+import * as Did from '@kiltprotocol/did'
+import { hexToU8a, stringToU8a, u8aToHex, u8aToString, isHex } from '@polkadot/util'
+
+/**
+   * Checks if delegation data is well formed.
+   *
+   * @param delegationData Delegation data to check.
+   */
+export function verifyDelegationStructure(delegationData: IDelegationData): void {
+  const { permissions, id, parentId, isPCR, account } = delegationData
+
+  if (!id) {
+    throw new SDKErrors.DelegationIdMissingError()
+  } else if (typeof id !== 'string' || !isHex(id)) {
+    throw new SDKErrors.DelegationIdTypeError()
+  }
+
+  if (!account) {
+    throw new SDKErrors.OwnerMissingError()
```

Comment:
what are your concerns?

Reply:
The reason they exist is so you can do instanceof checks, but these are actually incredibly brittle and fail easily. Not only in case we remove and replace them, but also in case the dependency tree gets messed up and there's dependency duplication in the tree.

Reply:
I introduced two new error types (for messaging and quoting). I completely eliminated the SDK error type.

Furthermore, I will submit a pull request in the SDK repository to remove certain types that are not necessary there.

[6bfd5d8](https://github.com/KILTprotocol/kilt-extension-api/pull/18/commits/6bfd5d8b6b6b03af30c3e82b372bf405c5294270) and [7670f94](https://github.com/KILTprotocol/kilt-extension-api/pull/18/commits/7670f94920517c1d8d3f30e790003aa945a32239)